File: app/src/main/java/com/amaury/pointage/v2/engine/CompanyEmployeeDeductionPayrollBridgeV2.kt
package com.amaury.pointage.v2.engine

import com.amaury.pointage.v2.model.DeductionV2
import java.time.YearMonth

/**
 * Convertit uniquement les retenues salariales qui diminuent réellement le net versé,
 * confirmées, datées et sourcées, en entrées du moteur paie.
 *
 * La part salariale de prévoyance fiscalement non déductible n'est pas une seconde retenue cash :
 * elle fait déjà partie de la prévoyance salariale réellement prélevée et sert uniquement à la
 * réintégration du net imposable. La convertir ici en [DeductionV2] la déduirait deux fois.
 *
 * Les montants qui modifient seulement une assiette fiscale/sociale ne sont donc volontairement pas
 * transformés en retenues sur le salaire versé. L'absence d'un enregistrement daté n'est jamais assimilée
 * à zéro : pour confirmer l'absence d'une retenue cash sur un mois, un montant explicite de 0 doit exister.
 */
object CompanyEmployeeDeductionPayrollBridgeV2 {
    private val directEmployeeDeductionKinds = setOf(
        CompanyEmployeeDeductionResolverV2.Kind.MUTUAL_EMPLOYEE,
        CompanyEmployeeDeductionResolverV2.Kind.PROVIDENT_EMPLOYEE,
        CompanyEmployeeDeductionResolverV2.Kind.TRANSPORT_EMPLOYEE
    )

    data class Result(
        val deductions: List<DeductionV2>,
        /** Vrai seulement si chaque retenue salariale cash directe possède une valeur datée confirmée pour le mois. */
        val confirmedEmployeeDeductionsComplete: Boolean,
        val warnings: List<String>,
        val traces: List<String>
    )

    fun resolve(
        snapshot: CompanyEmployeeDeductionResolverV2.Snapshot,
        period: YearMonth
    ): Result {
        val deductions = mutableListOf<DeductionV2>()
        val warnings = mutableListOf<String>()
        val traces = mutableListOf<String>()
        var complete = true

        directEmployeeDeductionKinds
            .sortedBy { it.ordinal }
            .forEach { kind ->
                val value = snapshot[kind]
                when {
                    !value.reliable -> {
                        complete = false
                        warnings += value.warnings.ifEmpty {
                            listOf("${kind.label} : valeur non fiable pour $period ; aucune retenue n'est inventée.")
                        }
                    }
                    value.amount == null -> {
                        complete = false
                        warnings += "${kind.label} : aucun montant daté confirmé pour $period ; aucune retenue n'est inventée."
                    }
                    else -> {
                        val amount = value.amount
                        if (!amount.isFinite() || amount < 0.0) {
                            complete = false
                            warnings += "${kind.label} : montant invalide pour $period ; aucune retenue n'est inventée."
                        } else {
                            if (amount > 0.0) {
                                deductions += DeductionV2(
                                    id = "company_employee_${kind.name.lowercase()}",
                                    label = kind.label,
                                    amount = amount,
                                    recurring = true
                                )
                            }
                            traces += buildString {
                                append(kind.label)
                                append(" : ")
                                append(amount)
                                append(" € confirmés pour ")
                                append(period)
                                value.source?.takeIf { it.isNotBlank() }?.let { append(" — source ").append(it) }
                            }
                        }
                    }
                }
            }

        return Result(
            deductions = deductions,
            confirmedEmployeeDeductionsComplete = complete,
            warnings = warnings.distinct(),
            traces = traces.distinct()
        )
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/CompanyEmployeeDeductionPayrollBridgeV2.swift
import Foundation

enum CompanyEmployeeDeductionPayrollBridgeV2 {
    struct Result: Equatable {
        let deductions: [DeductionV2]
        let confirmedEmployeeDeductionsComplete: Bool
        let warnings: [String]
        let traces: [String]
    }

    private static let directKinds: [CompanyEmployeeDeductionResolverV2.Kind] = [
        .mutualEmployee,
        .providentEmployee,
        .transportEmployee,
        .employeeProvidentNonDeductible
    ]

    static func resolve(
        snapshot: CompanyEmployeeDeductionResolverV2.Snapshot,
        period: CompanyEmployeeDeductionResolverV2.YearMonth
    ) -> Result {
        var deductions: [DeductionV2] = []
        var warnings: [String] = []
        var traces: [String] = []
        var complete = true

        for kind in directKinds {
            let value = snapshot[kind]
            if !value.reliable {
                complete = false
                warnings.append(contentsOf: value.warnings.isEmpty
                    ? ["\(kind.label) : valeur non fiable pour \(period) ; aucune retenue n'est inventée."]
                    : value.warnings)
                continue
            }

            guard let amount = value.amount else {
                complete = false
                warnings.append("\(kind.label) : aucun montant daté confirmé pour \(period) ; aucune retenue n'est inventée.")
                continue
            }

            guard amount.isFinite, amount >= 0 else {
                complete = false
                warnings.append("\(kind.label) : montant invalide pour \(period) ; aucune retenue n'est inventée.")
                continue
            }

            if amount > 0 {
                deductions.append(
                    DeductionV2(
                        id: "company_employee_\(kind.rawValue.lowercased())",
                        label: kind.label,
                        amount: amount,
                        recurring: true
                    )
                )
            }

            var trace = "\(kind.label) : \(amount) € confirmés pour \(period)"
            if let source = value.source?.trimmingCharacters(in: .whitespacesAndNewlines), !source.isEmpty {
                trace += " — source \(source)"
            }
            traces.append(trace)
        }

        return Result(
            deductions: deductions,
            confirmedEmployeeDeductionsComplete: complete,
            warnings: unique(warnings),
            traces: unique(traces)
        )
    }

    private static func unique(_ values: [String]) -> [String] {
        var seen = Set<String>()
        return values.filter { seen.insert($0).inserted }
    }
}
